File: backend/app/controller/article.js
'use strict';

const Controller = require('egg').Controller;

const util = require('../utils/commonUtil');

class ArticleController extends Controller {
  async query() {
    let ctx = this.ctx;
    let query = {
      'include': [{ model: this.app.model.Skill}]
    }

    let body = {};

    if(typeof(ctx.request.body)!='undefined' && ctx.request.body!=null){
      body = ctx.request.body;
    }
    //处理分页
    const {currentPage,pageSize} = body;
    if(typeof(currentPage)!= 'undefined' &&typeof(pageSize)!= 'undefined'){
      query.limit = pageSize;
      query.offset = pageSize * (currentPage - 1);
    }else{
      query.limit = 10;
      query.offset = 0;
    }
    //处理排序
    const {sorter} = body;
    if(typeof(sorter)!= 'undefined'){
      if(sorter.indexOf("skill") != -1){
        util.dealSorterRelative(query,this.app.model.Article.associations.skill,sorter);
      }else{
        util.dealSorter(query,sorter);
      }
      
    }

    //处理title
    const {title} = body;
    if(typeof(title)!= 'undefined' && title != null && title != ''){
      util.dealKeyLike(query,'title',title);
    }

    //处理platform
    const {platform} = body;
    if(typeof(platform)!= 'undefined' && platform != null && platform != ''){
      util.dealKeyEqual(query,'platform',platform);
    }

    //处理id
    const {id} = body;
    if(typeof(id)!= 'undefined' && id != null && id != ''){
      util.dealKeyEqual(query,'id',id);
    }

    //处理skill
    const {first,second,third} = body;
    if(typeof(first)!= 'undefined'&& first != null && first != ''){
      //获取所有skill
      let skill_query = {}
      //处理first
      util.dealKeyEqual(skill_query,'first',first);
      //处理second
      if(typeof(second)!= 'undefined'&&second!=null &&second!=''){
        util.dealKeyEqual(skill_query,'second',second);
      }
      //处理third
      if(typeof(third)!= 'undefined'&&third!=null &&third!=''){
        util.dealKeyEqual(skill_query,'third',third);
      }

      const skills= await ctx.service.skill.query(skill_query);
      const array=[];
      if(skills.rows.length>0){
        skills.rows.map((data)=>{
          array.unshift(data.id);
        });
      }
      //console.log(array);
      util.dealKeyEqual(query,'skillId',array);

    }
    ctx.body = await ctx.service.article.query(query);
  }

  // async show() {
  //   const ctx = this.ctx;
  //   ctx.body = await ctx.service.user.find(ctx.helper.parseInt(ctx.params.id));
  // }

  async add() {
    const ctx = this.ctx;
    const article = ctx.request.body;
    //查询skill
    const query={
      where:{
        first:article.first,
        second:article.second,
        third:article.third,
      }
    }
    let skills = await ctx.service.skill.query(query);
    if(skills.rows.length != 1){
      ctx.status = 510;
      ctx.body = '查询分类失败!';
    }else{
      article.skillId = skills.rows[0].id;
      const obj = await ctx.service.article.create(article);
      ctx.status = 201;
      ctx.body = obj;
    }
    
  }

  async update() {
    const ctx = this.ctx;
    const id = ctx.helper.parseInt(ctx.params.id);
    const article = ctx.request.body;
    //获取分类
    //查询skill
    const query={
      where:{
        first:article.first,
        second:article.second,
        third:article.third,
      }
    }
    let skills = await ctx.service.skill.query(query);
    if(skills.rows.length != 1){
      ctx.status = 510;
      ctx.body = '查询分类失败!';
    }else{
      article.skillId = skills.rows[0].id;

      ctx.body = await ctx.service.article.update({ id, updates: article });
      ctx.status = 201;
    }

  }

  async delete() {
    const ctx = this.ctx;
    const id = ctx.helper.parseInt(ctx.params.id);
    await ctx.service.article.del(id);
    ctx.status = 200;
    ctx.body={
      id:id
    };
  }

    //获取所有平台信息
  async distinctPlatform(){
    const ctx = this.ctx;
    //参数检验
    let body = ctx.request.body;
    if(typeof(body)=='undefined'){
      body={};
    }
    
    //分别查询
    let sql = 'SELECT DISTINCT platform FROM article ';
    ctx.body = await this.app.model.query(sql, {type : this.app.model.QueryTypes.SELECT});
    ctx.status = 200;
  }
}



module.exports = ArticleController;

File: backend-golang/src/router/article.go
package router

import (
	"backend-golang/src/global"
	"backend-golang/src/middleware"
	"backend-golang/src/model/po"
	"backend-golang/src/model/request"
	"backend-golang/src/model/response"
	"backend-golang/src/util"
	"database/sql"
	"github.com/gin-gonic/gin"
	"github.com/gin-gonic/gin/binding"
	"net/http"
	"reflect"
	"strconv"
)

func initArticleRoute(Router *gin.RouterGroup) {
	Router.Group("article").Use(middleware.Auth())
	{
		global.GroupApi.POST("article", articleAll)
		global.GroupApi.POST("article/add", articleAdd)
		global.GroupApi.POST("article/delete/:id", articleDelete)
		global.GroupApi.POST("article/update/:id", articleUpdate)
		global.GroupApi.POST("article/distinctPlatform", articleDistinctPlatform)
	}

}

// @Tags Article
// @Summary 查询文章
// @Produce  application/json
// @Param data body request.ArticleQueryRequest true "查询文章"
// @Success      200  {object}  response.Response
// @Router /api/v1/article [post]
func articleAll(c *gin.Context) {
	req := request.DefaultArticleQueryRequest()
	if reflect.TypeOf(c.Request.Body) != reflect.TypeOf(http.NoBody) {
		err := c.ShouldBindBodyWith(&req, binding.JSON)
		if util.HandleError(c, err) {
			return
		}
	}
	order := util.ParseSorter(req.Sorter)
	where := make(map[string]interface{})
	if len(req.Title) > 0 {
		where["title like ?"] = "%" + req.Title + "%"
	}
	if req.Id > 0 {
		where["article.id = ?"] = req.Id
	}
	if len(req.Platform) > 0 {
		where["platform = ?"] = req.Platform
	}
	if len(req.Type) > 0 && len(req.First) > 0 && len(req.Second) > 0 && len(req.Third) > 0 {
		classifyKey := po.NewClassificationByKey(req.ClassificationKey)
		classify, err := classifyKey.One()
		if err != nil || classify == nil {
			response.FailWithMessage("未查询到类型"+util.JsonNoException(classifyKey), c)
			return
		}
		where["classification = ?"] = classify.Id
	}

	query := &util.PaginationQuery{
		Limit:  req.PageSize,
		Offset: req.PageSize * (req.CurrentPage - 1),
		Order:  order,
		Where:  where,
	}
	list, total, err := (&po.ArticleWithClassification{}).AllWithClassification(query)
	list2 := make([]po.ArticleWithClassificationVO, 0)
	if list != nil && len(*list) > 0 {
		list2 = make([]po.ArticleWithClassificationVO, len(*list))
		for i, val := range *list {
			list2[i] = po.NewArticleWithClassification(val)
		}
	}

	if util.HandleError(c, err) {
		return
	}
	data := response.PageResponse{total, list2}
	util.JsonData(c, data)
}

// @Tags Article
// @Summary 新增文章
// @Produce  application/json
// @Param data body request.ArticleAddOrUpdateRequest true "x新增文章"
// @Success      200  {object}  response.Response
// @Router /api/v1/article/add [post]
func articleAdd(c *gin.Context) {
	var req request.ArticleAddOrUpdateRequest
	if reflect.TypeOf(c.Request.Body) != reflect.TypeOf(http.NoBody) {
		err := c.ShouldBindBodyWith(&req, binding.JSON)
		if util.HandleError(c, err) {
			return
		}
	}
	// 获取 classfication
	classifyKey := po.NewClassificationByKey(req.ClassificationKey)
	classify, err := classifyKey.One()
	if err != nil || classify == nil {
		response.FailWithMessage("未查询到类型"+util.JsonNoException(classifyKey), c)
	}
	req.Classification = int(classify.Id)
	article := po.NewArticle(req)
	err = article.CreateUserOfRole()
	if util.HandleError(c, err) {
		return
	}
	util.JsonData(c, article)
}

// @Tags Article
// @Summary 删除用户
// @Produce  application/json
// @Success      200  {object}  response.Response
// @Router /api/v1/article/delete/{id} [post]
func articleDelete(c *gin.Context) {
	var mdl po.Article
	id, err := util.ParseParamID(c)
	if util.HandleError(c, err) {
		return
	}
	mdl.Id = id
	err = mdl.Delete()
	if util.HandleError(c, err) {
		return
	}
	util.JsonData(c, mdl)
}

// @Tags Article
// @Summary 更新用户
// @Produce  application/json
// @Success      200  {object}  response.Response
// @Router /api/v1/article/update/{id} [post]
func articleUpdate(c *gin.Context) {
	var req request.ArticleAddOrUpdateRequest
	if reflect.TypeOf(c.Request.Body) != reflect.TypeOf(http.NoBody) {
		err := c.ShouldBindBodyWith(&req, binding.JSON)
		if util.HandleError(c, err) {
			return
		}
	}
	id, err := util.ParseParamID(c)
	if util.HandleError(c, err) {
		return
	}
	req.Id = id
	article := po.Article{}
	article.Id = id
	inDb, err := article.One()
	if util.HandleError(c, err) {
		return
	}
	if inDb == nil {
		response.FailWithMessage("未找到用户"+strconv.Itoa(int(req.Id)), c)
	}
	// 获取 classfication
	classifyKey := po.NewClassificationByKey(req.ClassificationKey)
	classify, err := classifyKey.One()
	if err != nil || classify == nil {
		response.FailWithMessage("未查询到类型"+util.JsonNoException(classifyKey), c)
	}
	article = po.NewArticle(req)
	article.Classification = int(classify.Id)
	article.Update()
	util.JsonData(c, article)
}

// @Tags Article
// @Summary 查询平台列表
// @Produce  application/json
// @Success      200  {object}  response.Response
// @Router /api/v1/article/distinctPlatform [post]
func articleDistinctPlatform(c *gin.Context) {
	sqlStr := "SELECT DISTINCT platform FROM article"
	rows, err := global.Db.Query(sqlStr)
	if util.HandleError(c, err) {
		return
	}
	ans := []string{}
	for rows.Next() {
		var val sql.NullString
		err = rows.Scan(&val)
		if err != nil {
			global.Logger.Error(err)
			continue
		}
		ans = append(ans, val.String)
	}
	util.JsonData(c, ans)
}
